Allow arbitrary "categorical value" types

alyst opened the pull request:
To transparently support string data, string categorical value type need to be inherited from `AbstractString`, which would conflict with the other categorical types (e.g. numeric) if all are implemented by the same `CategoricalValue{T, R}`.
Based on the discussions in #77, I put this very early draft of the changes that allow `AbstractString` inheritance and dispatch for the string categories, while preserving the old behavior for any other categorical types.
Please give your early feedback on the approach taken.

This PR allows `CategoricalArray` to use arbitrary categorical value type and introduces `CategoricalString{R} <: AbstractString` to be used by default for `String` categorical values (`CategoricalValue{T,R}` is used for the rest).
Both `CategoricalString` and `CategoricalValue` should support "categorical value" concept, which currently requires to "iscatvalue/pool/level/valtype/reftype()" methods.

Since `CategoricalValue` and `CategoricalString` are not relatives, `CatValue = Union{CategoricalValue, CategoricalString}` is introduced to reduce the dispatch rules.
This is not a very good design as `CatValue` definition would have to be updated for each new "categorical value" implementation, making external categorical value types not possible.
However, it should be possible to avoid `CatValue` by employing categorical value traits (dispatching by the result of `iscatvalue(T)`).

nalimilan commented on src/array.jl:
```diff
@@ -19,9 +19,11 @@ function reftype(sz::Int)
+reftype(::Type{<:CategoricalArray{T, N, R}}) where {T, N, R} = R
+reftype(A::T) where {T <: CategoricalArray} = reftype(T)
+
+catvalue_type(::Type{<:CategoricalArray{T,N,R,V,C}}) where {T,N,R,V,C} = C
```
These methods aren't used AFAICT.

nalimilan commented on src/array.jl:
```diff
@@ -141,18 +143,15 @@ CategoricalArray{T}(m::Int, n::Int; ordered=false) where {T} =
+function CategoricalArray{T, N, R}(dims::NTuple{N,Int};
+                                   ordered=false) where {T<:CatValue, N, R<:Integer}
+    V = unwrap_catvalue_type(T)
+    CategoricalArray{V, N, R}(zeros(R, dims), CategoricalPool{V, R, T}(ordered))
```
I think you can call `CategoricalArray{T, N, R}(dims, ordered=ordered)`. That way all constructors fall back to a single one, that makes it easier to understand how all this mess is organized.

nalimilan commented on src/nullablearray.jl:
```diff
@@ -3,21 +3,21 @@ import Base: convert, getindex, setindex!, similar, in
+function CategoricalArray{Union{T, Null}, N, R}(dims::NTuple{N,Int};
+                                                ordered=false) where {T<:CatValue, N, R<:Integer}
+    V = unwrap_catvalue_type(T)
+    CategoricalArray{Union{V, Null}, N, R}(zeros(R, dims), CategoricalPool{V, R, T}(ordered))
```
Here too better use `CategoricalArray{Union{V, Null}, N, R}(dims, ordered=ordered)`.

nalimilan commented on src/pool.jl:
```diff
@@ -80,6 +82,9 @@ function Base.show(io::IO, pool::CategoricalPool{T, R}) where {T, R}
     pool.ordered && print(io, " with ordered levels")
 end
 
+reftype(::Type{CategoricalPool{T,R}}) where {T,R} = R
```
Unused fonction? Better remove it then.

nalimilan commented on src/typedefs.jl:
```diff
@@ -2,9 +2,10 @@ const DefaultRefType = UInt32
-# See JuliaLang/julia#269
+# Type params:
+# * `T` type of categorized values
+# * `R` integral type for referncing category levels
```
"integer", "referencing"

nalimilan commented on src/typedefs.jl:
```diff
@@ -2,9 +2,10 @@ const DefaultRefType = UInt32
+# Type params:
+# * `T` type of categorized values
+# * `R` integral type for referncing category levels
+# * `V` categorical value type that implements "categorical value" concept
```
Instead of "concept", which isn't a standard term, I'd just say "one of the `CatValue` types", and enforce this in the signature with `<:CatValue`. That makes it clear that for now this is not a public extensible interface. If you want, call this "trait", as "concept" is pretty vague (same below).

alyst replied:
I'm using "concept" in C++1x sense, but you're right, probably it's better to say "categorical value trait".

In the final version of PR I would like to reduce the use of `CatValue` to `values.jl` only.
I think it's possible to make `CategoricalArray` dispatch based on `iscatvalue(C)` only without introducing any more complexity than it currently has.

nalimilan replied:
I'm afraid dispatching on a trait will be more complex than using `CatValue`, as you need to pass `iscatvalue(C)` as a parameter instead of just dispatching on the argument's types. It's fine to use `CatValue`, no need to make things more complex unless/until we have a good reason to.

nalimilan commented on src/CategoricalArrays.jl:
```diff
@@ -1,14 +1,14 @@
     export LevelsException
 
-    export categorical, compress, decompress, droplevels!, levels, levels!, isordered, ordered!
+    export categorical, compress, decompress, droplevels!, levels, levels!, isordered, ordered!, iscatvalue
```
Better not export this for now as `CatValue` isn't supposed to be extended as long as it's not a real trait.

alyst commented on src/typedefs.jl:
```diff
@@ -29,28 +30,51 @@ end
+struct CategoricalArray{T, N, R <: Integer, V, C, U} <: AbstractCategoricalArray{T, N, R, V, C, U}
     refs::Array{R, N}
-    pool::CategoricalPool{V, R, CategoricalValue{V, R}}
+    pool::CategoricalPool{V, R, C}
```
That would be nice, but `CategoricalPool{T,R,V}` is explicitly using `T` in `index`, `invindex` and `levels`.
`T` is indeed extractable from `V`, but I don't know how to void using it when defining `struct`.

alyst replied:
Probably it's possible to simplify `CategoricalPool` constructors by providing only `C`, but that might make it harder to create standalone pools.

nalimilan replied:
> T is indeed extractable from V, but I don't know how to void using it when defining struct.

What do you mean?

I wouldn't worry about how hard it is to create a `CategoricalPool`, currently isn't not even exported. We could create convenience constructors if we want.

alyst replied:
"avoid", if that caused confusion

By "dropping parameter V" have you meant from the `struct` or from the constructors?

nalimilan replied:
From the `struct`. You can provide convenience constructors which guess `V` from the input parameters if that's useful.

nalimilan replied:
Have you tried this?

alyst replied:
I'm not sure I understand what you envision to do.
`V` is used in multiple fields of the `struct CategoricalPool`, so either it stays as the type parameter of the `struct` or the `struct` layout is changed, although I'm not sure it's feasible.
If you actually propose to simplify the `CategoricalPool` ctors type parameters, I can look into that.

nalimilan replied:
Ah, too bad. Looks like we'll have to live with all these semi-redundant type parameters then...

Maybe just rename the type parameters so that `V` stands for "value" for consistency with `CategoricalPool`. You could replace `V` with `L` for "levels". Though if that causes too much churn we can do that later.

alyst commented on src/value.jl:
```diff
@@ -1,145 +1,194 @@
+unwrap_catvalue_type(::Type{T}) where {T} = T
+
+# default categorical value type for given non-null value type `T` and reference type `R`
+catvalue_type(::Type{T}, refType::Type{R}) where {T<:CatValue, R} =
```
The intention is to be agnostic about the nature of `T`.
We don't know what `T` is, but if it's using the same `refType`, we keep using it.
If `T` uses a different reftype, the new categorical value type is constructed.

nalimilan replied:
`T<:CatValue`, so we completely know what it is. And even if we didn't, it would be quite weird to preserve `T` when `reftype(T) === R`, but not preserve it in other cases. `R` is really a secondary parameter compared to `valtype(T)`, it shouldn't be allowed to determine the kind of `CatValue` you get.

alyst replied:
The idea is to try to keep "value type" `T`, so that `convert()` doesn't do unnecessary copies. The `valtype(T)` would be preserved anyway.

nalimilan replied:
I don't see why `convert` would make a copy since `catvalue_type` should return the same type as `T` in the end, right?

alyst replied:
Yes. Right now there's only `CategoricalString` and `CategoricalValue` that wrap distinct types (`AbstractString` and everything else), so the default `catvalue_type(valtype(T), R)` would match `T`.
But I hope all the types are resolved at compile time, so `reftype(T) === R` check, which might be relevant for the future, costs nothing.

nalimilan replied:
I agree it costs nothing in terms of computation, but it costs in terms of complexity, and I'm not convinced it's useful right now. Let's design the simplest system which works now, and only make it more general later if we feel the need for it. It doesn't make much sense to design a general system without using it -- anyway, if we don't test its general capabilities, there will always be an undetected failure somewhere.

alyst commented on src/value.jl:
```diff
@@ -1,145 +1,194 @@
+
+# creates categorical value for `level` from the `pool`
+# The result type is of type `C`, which may be different from `CategoricalValue{T,R}`
+function CategoricalValue(level::Integer, pool::CategoricalPool{T, R, C}) where {T, R, C}
```
Actually I was inclined to cal it `catvalue()` to avoid any connotations with constructors of specific type(s).

nalimilan replied:
But that's actually a kind of constructor, just like you can call `AbstractArray([1])`.

alyst commented on test/01_typedef.jl:
```diff
@@ -39,7 +39,7 @@ module TestTypeDef
         x = CategoricalValue(i, pool)
 
-        @test isa(x, CategoricalValue)
+        @test iscatvalue(x)
```
I agree that there should be more tests covering different categorical types.
But how much should there be emphasis on the actual implementation?
I suppose as long as the "black box" is doing the right thing (e.g. `AbstractString` methods are working as expected on string categories) it doesn't matter what actual types it uses.
That should reduce the burden of tests maintenance.

nalimilan replied:
Since 80% of methods are common to all `CatValue` types, we don't really have to replicate the test for `CategoricalValue` and `CategoricalString`. Indeed as long as the "black box" works it should be OK, but AFAIK we only test the `AbstractString` interface on `CategoricalString` separately, and not in the context of `CategoricalArray{String}` (or at least not systematically). So it would be good to test in a few situations that we indeed get a `CategoricalString` or a `CategoricalValue` where we expect it (i.e. when indexing an array constructed by a few different means). Luckily these tests already exist, they just need an additional line where appropriate.

alyst commented on test/11_array.jl:
```diff
@@ -4,8 +4,10 @@ using Base.Test
 
-for ordered in (false, true)
-    for R in (CategoricalArrays.DefaultRefType, UInt8, UInt, Int8, Int)
+@testset "array" begin
```
+1 it just happens that with `@testset` I have more informative diagnostic in case of test failure.

alyst commented:
@nalimilan Thanks for the review! There's another question: currently `CategoricalString` supports only `String` values, should it be parameterized on `T<:AbstractString`?

alyst commented:
There are failures on nightlies (`UndefVarError: T not defined` when `unwrap_cavalue_type(Null)` is called) at this line:
```julia
unwrap_catvalue_type(::Type{Union{T, Null}}) where {T<:CatValue} = Union{valtype(T), Null}
```

Should I try to fix it, or it's a `julia` bug? I expected that `unwrap_cavalue_type(Null)` is dispatched to
```julia
unwrap_catvalue_type(::Type{T}) where {T} = T
```

nalimilan commented:
> There's another question: currently CategoricalString supports only String values, should it be parameterized on T<:AbstractString?

Hmm, good question. I guess it's OK to force `String` as long as the constructor accepts other kinds of `AbstractString` and does the conversion automatically. Using a different string type doesn't sound too useful since `CategoricalArray` is designed for situations where the number of levels is small so the processing of string levels should be negligible compared with the cost of recoding all individual reference codes. And that's always an extension that can be added later if somebody complains. Finally, `CategoricalString{String, UInt32}` sounds a bit ridiculous... :-)

Apparently the "T not defined" error is expected if you pass `Null` to `unwrap_catvalue_type(::Type{Union{T, Null}})`, see https://github.com/JuliaLang/julia/issues/24124. You can fix it by defining `unwrap_catvalue_type(::Null) = Union{}`, but I wonder whether there are legitimate situations in which `Null` can be passed... When does that happen?

alyst commented on src/value.jl:
```diff
@@ -1,145 +1,194 @@
 
-Base.hash(x::CategoricalValue, h::UInt) = hash(index(x.pool)[x.level], h)
-
-function Base.isless(x::CategoricalValue{S}, y::CategoricalValue{T}) where {S, T}
```
In fact, this method is related to this PR since in this PR the comparisons could be done between different categorical value implementations.
So this method could have been kept, but to cover all cases a bunch of `CategoricalValue` vs `CategoricalString` comparisons have to be added as well.
And yet it would not cover all statically resolvable `pool(x) !== pool(y)` cases since the reference types also have to be taken into account.  

Luckily, the compiler seems to be smart enough to reduce the common `isless()` to just throwing an exception.
```
julia> z = CategoricalVector([1, 2, 3], ordered=true)
3-element CategoricalArrays.CategoricalArray{Int64,1,UInt32}:
 1
 2
 3

julia> y = CategoricalVector{String, UInt8}(["b", "a", "b"], ordered=true)
3-element CategoricalArrays.CategoricalArray{String,1,UInt8}:
 "b"
 "a"
 "b"

julia> x = CategoricalVector{String, UInt32}(["b", "a", "b"], ordered=true)
3-element CategoricalArrays.CategoricalArray{String,1,UInt32}:
 "b"
 "a"
 "b"

julia> @code_warntype Base.isless(x[1], y[1])
Variables:
  #self#::Base.#isless
  x::Any
  y::Any

Body:
  begin  # line 160:
      return (CategoricalArrays.throw)($(Expr(:new, :(Base.ArgumentError), "CategoricalValue objects with different pools cannot be tested for order")))::Union{} # line 162:
  end::Union{}

julia> @code_warntype Base.isless(x[1], z[1])
Variables:
  #self#::Base.#isless
  x::Any
  y::Any

Body:
  begin  # line 160:
      return (CategoricalArrays.throw)($(Expr(:new, :(Base.ArgumentError), "CategoricalValue objects with different pools cannot be tested for order")))::Union{} # line 162:
  end::Union{}

julia> @code_warntype Base.isless(y[1], z[1])
Variables:
  #self#::Base.#isless
  x::Any
  y::Any

Body:
  begin  # line 160:
      return (CategoricalArrays.throw)($(Expr(:new, :(Base.ArgumentError), "CategoricalValue objects with different pools cannot be tested for order")))::Union{} # line 162:
  end::Union{}

julia> @code_warntype Base.isless(y[1], y[2])
Variables:
  #self#::Base.#isless
  x::CategoricalArrays.CategoricalString{UInt8}
  y::CategoricalArrays.CategoricalString{UInt8}

Body:
  begin 
      unless (Core.Intrinsics.not_int)(((Core.getfield)(x::CategoricalArrays.CategoricalString{UInt8}, :pool)::CategoricalArrays.CategoricalPool{String,UInt8,CategoricalArrays.CategoricalString{UInt8}} === (Core.getfield)(y::CategoricalArrays.CategoricalString{UInt8}, :pool)::CategoricalArrays.CategoricalPool{String,UInt8,CategoricalArrays.CategoricalString{UInt8}})::Bool)::Bool goto 4 # line 160:
      return (CategoricalArrays.throw)($(Expr(:new, :(Base.ArgumentError), "CategoricalValue objects with different pools cannot be tested for order")))::Union{}
      4:  # line 162:
      $(Expr(:inbounds, false))
      # meta: location /home/astukalov/.julia/v0.6/CategoricalArrays/src/value.jl order 52
      SSAValue(3) = (Core.getfield)((Core.getfield)(x::CategoricalArrays.CategoricalString{UInt8}, :pool)::CategoricalArrays.CategoricalPool{String,UInt8,CategoricalArrays.CategoricalString{UInt8}}, :order)::Array{UInt8,1}
      # meta: location multidimensional.jl getindex 245
      SSAValue(5) = (Base.zext_int)(Int64, (Core.getfield)(x::CategoricalArrays.CategoricalString{UInt8}, :level)::UInt8)::Int64
      # meta: pop location
      # meta: pop location
      $(Expr(:inbounds, :pop))
      SSAValue(6) = (Base.arrayref)(SSAValue(3), SSAValue(5))::UInt8
      $(Expr(:inbounds, false))
      # meta: location /home/astukalov/.julia/v0.6/CategoricalArrays/src/value.jl order 52
      SSAValue(0) = (Core.getfield)((Core.getfield)(y::CategoricalArrays.CategoricalString{UInt8}, :pool)::CategoricalArrays.CategoricalPool{String,UInt8,CategoricalArrays.CategoricalString{UInt8}}, :order)::Array{UInt8,1}
      # meta: location multidimensional.jl getindex 245
      SSAValue(2) = (Base.zext_int)(Int64, (Core.getfield)(y::CategoricalArrays.CategoricalString{UInt8}, :level)::UInt8)::Int64
      # meta: pop location
      # meta: pop location
      $(Expr(:inbounds, :pop))
      return (Base.ult_int)(SSAValue(6), (Base.arrayref)(SSAValue(0), SSAValue(2))::UInt8)::Bool
  end::Bool
```

nalimilan replied:
OK, good to know the compiler is that smart!

alyst commented:
Next round
 *  I've tried to address most of the review comments except the tests
 * made `iscatvalue()` return a type and implemented trait-based dispatch for `[unwrap_ ]catvalue_type()`. That allowed to remove `CatValue` dependency from `CategoricalArray` ctors. Ctors in "nullablearray.jl" are not required anymore too. So external categorical values do not seem a very complicated thing now.
 * The value type of `CategoricalString` is `String`, but the `eltype()` of the original array could be any `<: AbstractString`. That introduces a case where `valtype(<categorical value>)` differs from the element type of the original array.
 * `CategoricalValue` is renamed to `catvalue()`. I try to constrain the use of `CatValue` type to `value.jl` because I feel it provokes bad design decisions. Especially the concept of `CatValue` should not leak outside to give an impression that it's a proper type.

There is a pair of FIXMEs/TODOs to make decision/comment.

So at this stage we should be able to finalize the changes to the source code and add missing tests.

nalimilan commented on src/CategoricalArrays.jl:
```diff
@@ -1,13 +1,14 @@
            NullableCategoricalArray, NullableCategoricalVector, NullableCategoricalMatrix
     export LevelsException
 
+    export catvalue#, iscatvalue
```
?

alyst replied:
`catvalue` is the function that takes the place of `CategoricalValue` *function*
For `iscatvalue()` you can decide whether to export it or not or keep it as a remainder to export in the future.

nalimilan replied:
Better not export any of these (and not leave a comment). Anyway `catvalue` only works if you access the pool, which isn't exported currently either.

alyst replied:
OK, `CategoricalValue()` method "leaked" with the `CategoricalValue` type and was used in the tests, so I'll remove the exports and fix the tests to use `CategoricalArrays.catvalue()`.

nalimilan commented on src/array.jl:
```diff
@@ -117,10 +113,16 @@ function CategoricalMatrix end
+    C = catvalue_type(unwrap_catvalue_type(T), R)
+    V = valtype(C)
+    S = T >: Null ? Union{V, Null} : V
+    CategoricalArray{S, N}(zeros(R, dims), CategoricalPool{V, R, C}(ordered))
```
I don't understand what this does. It doesn't sound correct to create a `CategoricalArray{S, N}` when the caller explicitly requested a `CategoricalArray{T, N, R}`. Transformations of type parameters should only happen when they are obtained implicitly from the types of the arguments.

The job of unwrapping `CategoricalValue` eltypes should rather be done by higher-level constructors as was the case before.

alyst replied:
So what it exactly does:
 * `unwrap_catvalue_type(T)` gets the original type if `T` is categorical value
 * `C = catvalue_type()` selects the appropriate categorical value implementation for the original type
 * `V` is the actual value type that `C` encodes, `V` might be different from `Nulls.T(unwrap_catvalue_type(T))`, e.g. when `T == SubString`, but `convert(V, x::T)` should always be possible. As I noted in the comment the possibility `V != Nulls.T(unwrap_catvalue_type(T))` was introduced to handle `AbstractString` values.
 * `S` restores the nullability

It's true that it might be better to do type logic somewhere upstream.
But the same critics applies to ctors like `CategoricalArray{CategoricalValue{T}, N, R}` from the original code. So to keep the already substantial changes minimal I would rather keep it.

Essentially, before the PR the logic of type wrapping/transformations and handling nulls was defined by the "higher-level" ctors.
With `CategoricalValue` being split into multiple types it becomes harder to continue to do it this way. And these ctors still require the overhead of passing the normal parameters so it becomes harder to read and maintain.
So in this PR the logic of type wrapping/unwrapping is delegated to trait-based `unwrap_catvalue_type(T)`/`catvalue_type()` methods, and the "central" ctor that we are now discussing is the only place where these methods are called.
IMO it leads to leaner code.

nalimilan replied:
Hmm, let me think about this. That need to special-case `CatValue` wrapping a `CatValue` is really not pretty in the current code either.

nalimilan replied:
OK, I guess it's fine, especially if it allows removing higher-level constructors. I'm not sure I wrote it in the best way in the first place since I remember it was quite messy.

nalimilan commented on src/array.jl:
```diff
@@ -173,24 +163,22 @@ CategoricalMatrix{T}(m::Int, n::Int; ordered=false) where {T} =
+    res = convert(CategoricalArray{V, N, R}, A)
     if res.pool === A.pool # convert() only makes a copy when necessary
-        res = CategoricalArray{U, N, R}(res.refs, deepcopy(res.pool))
+        res = CategoricalArray{V, N}(res.refs, deepcopy(res.pool))
```
Why remove R? That was more explicit. Same below.

alyst replied:
Because it would require introducing `reftype(CategoricalPool)` that you wanted to remove.
`R` is already passed into inner ctor by `refs` and by `pool`.
Also in many cases inner ctor usage `R` was not explicitly available, so it was extracted from the `refs`.
This scheme seemed a little bit redundant for me, also the removal of `R` seemed to be in line with simplifying the `CategoricalPool` ctor params.

nalimilan replied:
In the present case `R` is available so better pass it. That also makes it certain and clear that `res` doesn't change type depending on whether pools are equal, which would be a serious problem.

alyst replied:
There's only want ctor that accepts `CategoricalPool` as the parameter: the inner one.
So all these cases with `R` removal are referring to it.

I can introduce a "low-level" outer ctor with exactly the same signature as the inner one + `R` in the type parameters list to make the cases like the one here this more explicit and do a triple `R` check.

nalimilan replied:
Ah, sorry, now I realize that you removed the `R` parameter to the inner constructor, which explains everything.

alyst replied:
So it's fine to keep it as it is now, or you prefer that I implement my suggestion from the comment above?

nalimilan replied:
No, let's say it's fine as-is, we have enough type parameters floating around already.

nalimilan commented on src/array.jl:
```diff
@@ -117,10 +113,16 @@ function CategoricalMatrix end
+
 CategoricalArray{T, N}(dims::NTuple{N,Int}; ordered=false) where {T, N} =
-    CategoricalArray{T, N, DefaultRefType}(dims, ordered=ordered)
+    CategoricalArray{T, N, reftype(T)}(dims, ordered=ordered)
```
Same remark here and below: I don't think we should allow `T<:CatValue` in these constructors as that doesn't reflect the actual type which will be returned.

alyst replied:
But with the master it is possible:
```
julia> x = categorical(["a", "b", "c"])
3-element CategoricalArrays.CategoricalArray{String,1,UInt32}:
 "a"
 "b"
 "c"

julia> eltype(x)
CategoricalArrays.CategoricalValue{String,UInt32}

julia> CategoricalArray{eltype(x), 2}((2,2))
2×2 CategoricalArrays.CategoricalArray{String,2,UInt32}:
 #undef  #undef
 #undef  #undef
```

nalimilan replied:
Right, that comment is outdated given my reply above at https://github.com/JuliaData/CategoricalArrays.jl/pull/87#discussion_r145963843.

nalimilan commented on src/pool.jl:
```diff
@@ -3,39 +3,46 @@ function CategoricalPool(index::Vector{S},
     invindex = convert(Dict{S, R}, invindex)
-    CategoricalPool{S, R, CategoricalValue{S, R}}(index, invindex, order, ordered)
+    C = catvalue_type(S, R)
+    V = valtype(C) # might be diffrent from S (e.g. S==SubString, V==String)
```
"different". Also spaces around `==`.

nalimilan commented on src/value.jl:
```diff
@@ -1,145 +1,181 @@
+
+# "categorical value" trait implementation for CategoricalValue and CategoricalString
+iscatvalue(::Type{<:CatValue}) = IsCatValue
+valtype(::Type) = Union{} # no value type if not a proper categorical value type
```
Is this actually used?

alyst replied:
Yes, AFAIR it gets called for `categorical([null, null])` test.

nalimilan replied:
Is `valtype(Null)` called then? Or `valtype(Union{})`? Maybe we should better define another method to catch these cases before they are passed to `valtype`? Or just restrict this to `valtype(::Null)` or `valtype(::Union{})`? Defining this method for any type sounds weird.

alyst replied:
It turns out that I was wrong and in the current version we don't need this method, so I will ditch it for good.

nalimilan commented on src/value.jl:
```diff
@@ -1,145 +1,181 @@
+
+# extract the type of the original value from array eltype `T`
+unwrap_catvalue_type(::Type{T}) where {T} =
+    unwrap_catvalue_type(iscatvalue(T), T)
```
Using `iscatvalue(T)` here is really overkill when you could just dispatch on `<:CatValue`.

nalimilan commented on src/value.jl:
```diff
@@ -1,145 +1,181 @@
+end
+
+catvalue_type(::Type{IsCatValue}, ::Type{T}, ::Type{R}) where {T, R} =
+    reftype(T) === R ? T : catvalue_type(valtype(T), R)
```
I'm still not convinced the check is useful. Cf. https://github.com/JuliaData/CategoricalArrays.jl/pull/87#discussion_r145643980.

nalimilan commented on src/value.jl:
```diff
@@ -1,145 +1,181 @@
+Base.convert(::Type{T}, x::T) where {T <: CatValue} = x
+Base.convert(::Type{Union{T, Null}}, x::T) where {T <: CatValue} = x
+
+# FIXME do we need this rule or promotion is only required for CategoricalString?
```
If there aren't any test failures, we can assume it's enough. I think the ambiguities come from the fact that `CategoricalString <: AbstractString`.

alyst replied:
This line is needed to pass tests in `05_convert.jl`, e.g.
```julia
pool = CategoricalPool([1, 2, 3])
v1 = catvalue(1, pool)
@test promote(1, v1) === (1, 1)
```
So if you want to preserve this behaviour for non-string factors, we should keep it.
The other test sets don't depend on it.

nalimilan replied:
Yes, I think we want to preserve this behavior. I thought your comment meant that tests passed without it.

nalimilan commented on src/value.jl:
```diff
@@ -1,145 +1,181 @@
-Base.matchall(r::Regex, s::CategoricalValue{<:AbstractString}, overlap::Bool=false) =
+Base.matchall(r::Regex, s::CategoricalString, overlap::Bool=false) =
     matchall(r, get(s), overlap)
+Base.collect(x::CategoricalString) = collect(get(x))
```
Don't forget to test this (or leave for another PR if that's not really required for this one).

nalimilan replied:
Done.

nalimilan commented on src/typedefs.jl:
```diff
@@ -29,28 +33,48 @@ end
 
-struct CategoricalValue{T, R <: Integer} <: AbstractString
+"""
+Default implementation of "categorical value" concept
```
Still a few occurrences of "concept"?

alyst replied:
oops, I thought I did replace all

nalimilan commented on src/recode.jl:
```diff
@@ -294,16 +294,16 @@ julia> using CategoricalArrays, Nulls
 
 julia> recode(1:10, 1=>100, 2:4=>0, [5; 9:10]=>-1, 6=>null)
 10-element Array{Union{Int64, Nulls.Null},1}:
- 100    
```
Unrelated.

nalimilan commented:
Thanks.

> made iscatvalue() return a type and implemented trait-based dispatch for [unwrap_ ]catvalue_type(). That allowed to remove CatValue dependency from CategoricalArray ctors. Ctors in "nullablearray.jl" are not required anymore too. So external categorical values do not seem a very complicated thing now.

Honestly this still looks like overengineering to me. `iscatvalue` is only used once outside of value.jl (in the `CategoricalPool` inner constructor). That doesn't justify creating a full trait machinery. It's even misleading to use a trait in this case since most methods are defined on `CatValue`, so an external implementation would have to rewrite all of them. If you really don't want to use `CatValue` outside value.jl, just leave `V` unrestricted (duck typing): `CategoricalPool` is not exported and it's no like people will pass any type.

alyst commented:
> Honestly this still looks like overengineering to me. `iscatvalue` is only used once outside of value.jl (in the `CategoricalPool` inner constructor). 

But that's exactly the point, the use of `iscatvalue` is confined to just a few lines of code (`[unwrap_]catvalue_type()` methods).
It handles the logic that previously was encoded by the `CategoricalArray` ctors "family" and allows to reduce the number of ctors while supporting more flexible categorical value type system.
The use in `CategoricalPool` is optional and could be removed, although it's nice to have this type check.
True, it could also be implemented with `CatValue`, but that would require no less code and no less complexity.

To really support "categorical value" trait, the class has to implement a few methods (`valtype()`, `get()`,`pool()`,`level()`,`promote_rule()`).
`CatValue` is exclusively used to simplify the definition of these methods for `CategoricalValue{T}` and `CategoricalString`.

nalimilan commented:
> True, it could also be implemented with CatValue, but that would require no less code and no less complexity.

Yes, it would be simpler since there would be one concept less to introduce (one function and one type) and you wouldn't have to define methods which pass the trait as the first argument for dispatch. Let's keep this simple for now, extending it later will be easy if it's really needed. But currently I don't see a use case for an external package to define custom categorical values.

nalimilan commented on src/typedefs.jl:
```diff
@@ -17,6 +18,9 @@ mutable struct CategoricalPool{T, R <: Integer, V}
                                       invindex::Dict{T, R},
                                       order::Vector{R},
                                       ordered::Bool) where {T, R, V}
+        if iscatvalue(V) !== IsCatValue
```
It would make sense to check that `reftype(V) === R`, else we could end up in a strange situation (even if in practice things might work).

nalimilan commented on src/array.jl:
```diff
@@ -141,18 +143,6 @@ CategoricalArray{T}(m::Int, n::Int; ordered=false) where {T} =
-CategoricalArray{CategoricalValue{T}, N}(dims::NTuple{N,Int};
-                                         ordered=false) where {T, N} =
-    CategoricalArray{T, N}(dims, ordered=ordered)
 #(::Type{CategoricalArray{CategoricalValue, N}}){N}(dims::NTuple{N,Int}; ordered=false) =
```
Remove these if you remove the similar ones above.

nalimilan commented on src/array.jl:
```diff
@@ -236,26 +224,27 @@ CategoricalMatrix(A::CategoricalArray{T, 2, R};
 convert(::Type{CategoricalArray}, A::AbstractArray{T, N}) where {T, N} =
     convert(CategoricalArray{T, N}, A)
 
+# FIXME T=String would use wrong categorical value, also it's not extensible, where it's used?
```
Are these methods still needed since the constructor above already ensured we don't wrap `CategoricalValue` inside `CategoricalValue`? Removing them would fix the problem. :-)

nalimilan commented on src/array.jl:
```diff
@@ -500,9 +489,10 @@ function vcat(A::CategoricalArray...)
     T = Base.promote_eltype(A...) >: Null ?
         Union{eltype(newlevels), Null} : eltype(newlevels)
-    refs = DefaultRefType[refsvec...;]
+    R = reftype(T)
```
AFAICT `T` cannot be a categorical value here, so no need to call `reftype`: just keep using `DefaultRefType`.

nalimilan commented on src/pool.jl:
```diff
@@ -3,39 +3,46 @@ function CategoricalPool(index::Vector{S},
     CategoricalPool(T[], Dict{T, R}(), R[], ordered)
 CategoricalPool{T}(ordered::Bool=false) where {T} =
-    CategoricalPool(T[], Dict{T, DefaultRefType}(), DefaultRefType[], ordered)
+    CategoricalPool{T, reftype(T)}(ordered)
```
Do we really need to call `reftype` here and below? I would assume `T` cannot be a categorical value, as `CategoricalArray` constructors should have controlled the parameters already?

alyst replied:
`reftype(T)` returns the default reference type for non-"categorical value" types. I.e. `DefaultRefType`.
It could be extended to return e.g. `UInt8` for `UInt8` or from the user code for custom types.

Some of these ctors are used in `recode.jl` or `extras.jl` (outside of `CategoricalArray`), some are just tested.
I can look into removing `CategoricalPool` ctor that are only used from tests.
I will add the check that `V` cannot be a categorical value type into the inner ctor of `CategoricalPool` (plus `reftype(C) === R` and `valtype(C) === V` checks).

nalimilan replied:
> reftype(T) returns the default reference type for non-"categorical value" types. I.e. DefaultRefType.
> It could be extended to return e.g. UInt8 for UInt8 or from the user code for custom types.

As I noted before, please implement the simplest system for now and leave possible extensions by external packages for later. So here let's just keep `DefaultRefType`, the diff is already large enough.

Indeed these constructors existed before `CategoricalArray`, so some of them may not be useful and could be removed in another PR.

alyst replied:
I have considered reverting `reftype(T)` to `DefaultRefType`, but then realized it would result in resetting of reference type in constructors like `CategoricalArray(x::AbstractArray)`, where `x` could actually be `CategoricalArray`.
I don't think it would be the expected behaviour, so I'll keep the generic `reftype(T)`.

nalimilan replied:
But we're dealing with the `CategoricalPool` constructor here, not the `CategoricalArray` one. If `T` is a categorical value, we're screwed anyway since there's no code in the constructors to unwrap `T` -- and indeed you just added checks to throw an error in that situation.

alyst commented:
To move on, I've replaced trait-based catvalue dispatch with `<:CatValue` based. I wouldn't call the former a good example of overdesigning as after the change it's still 9 methods. I'll put it into separate PR to save my work.

Also I've extracted whitespace fixes and @testsets. I'll put that into separate PRs as well.

alyst commented:
Should we deprecate `CategoricalValue` (and do not export `CategoricalString`) given that now `CategoricalValue` should serve less common level types, and the other packages should rather rely on `catvalue_type()` mechanism to get the proper categorical value type?

There are few uses of `CategoricalValue` in `DataFrames`.
In particular there's: `elty <: CategoricalValue`. It could be replaced by `<: CatValue`, but I don't like leaking `CatValue` into the wild.
`iscatvalue(elty)` seems a better solution.

alyst commented:
Now I recall that `valtype(Union{})` was there to handle 0.7 failure (`catvalue_type(Nulls.T(Null), )` call was dispatched to a more specific `T <: CatValue` method).
The current fix should be nicer.

nalimilan commented on src/value.jl:
```diff
@@ -35,6 +35,9 @@ catvalue_type(::Type{T}, ::Type{R}) where {T, R} =
 catvalue_type(::Type{<:AbstractString}, ::Type{R}) where {R} =
     CategoricalString{R}
+# default to CategoricalString in degenerated case (all nulls)
+catvalue_type(::Type{Null}, ::Type{R}) where {R} =
```
This sounds a bit weird. Doesn't it work if you return `CategoricalValue{Union{}, R}`? That's really the correct type, since we know it won't contain anything.

nalimilan commented:
> Should we deprecate CategoricalValue (and do not export CategoricalString) given that now CategoricalValue should serve less common level types, and the other packages should rather rely on catvalue_type() mechanism to get the proper categorical value type?
> 
> There are few uses of CategoricalValue in DataFrames.
> In particular there's: elty <: CategoricalValue. It could be replaced by <: CatValue, but I don't like leaking CatValue into the wild.
> iscatvalue(elty) seems a better solution.

At this stage I don't think there are lots of packages which rely on the internals of CategoricalArrays. Probably DataFrames and RCall, that's all. So we can handle this without deprecations. `CategoricalString` should be exported since it's exposed to users via `getindex`.

I don't have any problem with using `<: CatValue` in DataFrames, but let's discuss that later in your traits PR.

nalimilan commented on src/typedefs.jl:
```diff
@@ -29,28 +42,48 @@ end
 
-struct CategoricalValue{T, R <: Integer} <: AbstractString
+"""
+Default "categorical value" type for
```
Could you remove quotes around mentions of "categorical value"? I don't find it very useful, and for example in the docstring below they are not present.

nalimilan commented on src/value.jl:
```diff
@@ -1,145 +1,167 @@
+
+# checks whether the type is "categorical value"
+iscatvalue(::Type) = false
+iscatvalue(::Type{Union{}}) = false # otherwise it dispatches to Type{<:...}
```
I don't understand the comment. " dispatches to Type{<:...}" seems to imply it would dispatch to the method above, which sounds fine. Maybe this could be expanded a bit?

alyst replied:
Sorry, somehow I've missed that last round of comments. I will take a look shortly.

nalimilan commented on src/value.jl:
```diff
@@ -1,145 +1,167 @@
+unwrap_catvalue_type(::Type{T}) where {T} = T
+unwrap_catvalue_type(::Type{T}) where {T >: Null} =
+    Union{unwrap_catvalue_type(Nulls.T(T)), Null}
+unwrap_catvalue_type(::Type{Any}) = Any # to prevent dispatching to T>:Null method
```
Is that really a problem? `Union{Any, Null} == Any`. Same question for the `catvalue_type` method below.

nalimilan commented on src/value.jl:
```diff
@@ -1,145 +1,167 @@
+catvalue_type(::Type{T}, ::Type{R}) where {T >: Null, R} =
+    catvalue_type(Nulls.T(T), R)
+catvalue_type(::Type{T}, ::Type{R}) where {T <: CatValue, R} =
+    reftype(T) === R ? T : catvalue_type(valtype(T), R)
```
Still the same  remark about `reftype(T) === R`.

nalimilan commented on src/value.jl:
```diff
@@ -1,145 +1,167 @@
+Base.convert(::Type{T}, x::T) where {T <: CatValue} = x
+Base.convert(::Type{Union{T, Null}}, x::T) where {T <: CatValue} = x
+
+# FIXME do we need this rule or promotion is only required for CategoricalString?
```
Remove the FIXME?

nalimilan commented on test/01_typedef.jl:
```diff
@@ -37,15 +41,17 @@ module TestTypeDef
-        @test isa(x, CategoricalValue)
+        @test iscatvalue(x)
+        @test valtype(typeof(x)) == String
+        @test reftype(typeof(x)) == DefaultRefType
```
Please also test directly `x isa CategoricalValue` or even  `x isa CategoricalValue{Int, DefaultRefType}`, that doesn't hurt. Same below.

nalimilan commented on test/01_typedef.jl:
```diff
@@ -37,15 +41,17 @@ module TestTypeDef
-        @test isa(x.pool, CategoricalPool)
-        @test x.pool === pool
+        @test isa(CategoricalArrays.pool(x), CategoricalPool)
+        @test CategoricalArrays.pool(x) === pool
```
Add `pool` to `using` list since you did that for other functions.

alyst commented:
> I don't have any problem with using <: CatValue in DataFrames, but let's discuss that later in your traits PR.

It's not necessarily linked to traits as `iscatvalue()` or any other better-looking alternative could be implemented without it.

nalimilan commented:
> It's not necessarily linked to traits as iscatvalue() or any other better-looking alternative could be implemented without it.

Yeah, but it wouldn't be type-stable, which could be a problem in some cases (maybe not for DataFrames, though). So this needs discussing.

nalimilan commented:
> Now I recall that valtype(Union{}) was there to handle 0.7 failure (catvalue_type(Nulls.T(Null), ) call was dispatched to a more specific T <: CatValue method).
> The current fix should be nicer.

There's still an error on 0.7 though, so maybe reintroduce it anyway.

Also I forgot, but could you add `isa` tests to the array files as I mentioned before?

alyst commented:
> There's still an error on 0.7 though, so maybe reintroduce it anyway.

0.7 passes tests now, `unwrap_catvalue_type()` required similar fix.

alyst commented:
> How could `Null` dispatch to `::Type{<: CatValue}`?

`Nulls.T(Null)` gives `Union{} <: CatValue`

nalimilan commented:
> Nulls.T(Null) gives Union{} <: CatValue

Ah, right. Would be good to make the comment more explicit then, since it escaped me even though I'm the one who changed `Nulls.T` recently to return `Union{}`.

alyst commented:
> Also I forgot, but could you add isa tests to the array files as I mentioned before?

I've added more checks for the types of categorical values.
Actually, I think it's better not to check the exact type of the categorical value directly, but whether it uses the correct value and reference types (`valtype()` and `reftype()`).
The same for the other categorical array and pool types. This reduces the tests and dependent packages dependency on the actual implementation.
That's why I wanted to define `reftype()`, `valtype()` etc for array and pool.

Re `valtype()`: actually, there's `Base.valtype()` for getting the type of values in the associative collection.
Should we override `Base.valtype(CatValue)` or better change the name?
We don't need to export `valtype()` right now, but it might be required to reduce the aforementioned dependencies on implementation, so it's better to avoid name clash now.
`catvalue_type()` probably would have to be exported too, so maybe rename it into something more standard (`catvaltype()`?)?

While testing categorical values for `Date`, I've discovered that `CategoricalPool.levels` were constructed as `index[order]`, while it should be the other way around `index = levels[order]`.
I've committed a fix, but probably you would like to cherry-pick it into the new PR and add more thorough tests. I can rebase this PR after.

nalimilan commented:
> Actually, I think it's better not to check the exact type of the categorical value directly, but whether it uses the correct value and reference types (valtype() and reftype()).
> The same for the other categorical array and pool types. This reduces the tests and dependent packages dependency on the actual implementation.
> That's why I wanted to define reftype(), valtype() etc for array and pool.

But we're not an external package. We want to test exactly the behavior we expect in addition to the generic API -- tests can be changed later if we change the implementation. If we don't test that indexing into a `CategoricalArray{String}` gives a `CategoricalString`, we'd have to test that *all* `AbstractString` methods work, which isn't realistic. If you only test `valtype`, `CategoricalValue{String}` would pass the test, yet the `AbstractString` interface wouldn't be supported.

> Re valtype(): actually, there's Base.valtype() for getting the type of values in the associative collection.
> Should we override Base.valtype(CatValue) or better change the name?
> We don't need to export valtype() right now, but it might be required to reduce the aforementioned dependencies on implementation, so it's better to avoid name clash now.
> catvalue_type() probably would have to be exported too, so maybe rename it into something more standard (catvaltype()?)?

Ah, good catch, that's not ideal. `eltype` would be quite natural. Exporting it would indeed be useful since people can't just extract the `T` parameter now.

> While testing categorical values for Date, I've discovered that CategoricalPool.levels were constructed as index[order], while it should be the other way around index = levels[order].
> I've committed a fix, but probably you would like to cherry-pick it into the new PR and add more thorough tests. I can rebase this PR after.

Hmm, that sounds bad, but I wonder how it could have gone unnoticed until now. Yes, please file another PR, with a test which would previously fail. You can also move the commit adding comments there, since it's unrelated to the present PR.

alyst commented:
> If we don't test that indexing into a `CategoricalArray{String}` gives a `CategoricalString`, we'd have to test that all `AbstractString` methods work, which isn't realistic.

Sure, the simplicity of the tests should be the first priority.
Though in this particular case one way to test that string categories are properly supported is to create a `CategoricalArray{String}`, check that `valtype() == String`, that `catvaltype() <: AbstractString` and that all expected `AbstractString` operations work. That would essentially test the correctness of `CategoricalString` implementation without ever directly referring to it.

> Ah, good catch, that's not ideal. eltype would be quite natural. Exporting it would indeed be useful since people can't just extract the T parameter now.

`eltype()` might be good if applied to `CategoricalValue` only, but it would be nice if the same method also works for `CategoricalArray` (which has its own `eltype()` already) and `CategoricalPool`.
How about `lvltype()` or `leveltype()`?
How would you call the method that returns the "categorical value" type, i.e. 3rd param of `CategoricalPool` and similar contexts? Right now it's called `catvalue_type()`.

> Yes, please file another PR, with a test which would previously fail.

Done

nalimilan commented:
> Though in this particular case one way to test that string categories are properly supported is to create a CategoricalArray{String}, check that valtype() == String, that catvaltype() <: AbstractString and that all expected AbstractString operations work. That would essentially test the correctness of CategoricalString implementation without ever directly referring to it.

Yes, but really what's your problem with explicitly referring to these types? You seem to be acting as if we were in an external package and that you were afraid of possible breakage due to changes to internal functions. But we're testing our own code, it's fine to rely on internals! :-)

> eltype() might be good if applied to CategoricalValue only, but it would be nice if the same method also works for CategoricalArray (which has its own eltype() already) and CategoricalPool.
> How about lvltype() or leveltype()?

`leveltype` would be fine, but then I wouldn't export it as it doesn't seem worth extending the API without clear indications that it's needed.

> How would you call the method that returns the "categorical value" type, i.e. 3rd param of CategoricalPool and similar contexts? Right now it's called catvalue_type().

I don't have a problem with `catvalue_type`, let's just make sure we're consistent with `leveltype` regarding the use of underscores.

alyst commented:
> Yes, but really what's your problem with explicitly referring to these types? You seem to be acting as if we were in an external package and that you were afraid of possible breakage due to changes to internal functions.

Ideally the tests shouldn't be touched as long as there's no API/behaviour change. It's also a good test of API completeness. Which brings us to:

> `leveltype` would be fine, but then I wouldn't export it as it doesn't seem worth extending the API without clear indications that it's needed.

That's the way to fetch `T` from `CategoricalValue{T,R}`, it was a minor problem when there was just `CategoricalValue`, but with an addition of `CategoricalString` there should be a uniform way to get it from any categorical value [type]. If you don't know `T`, it makes it e.g. harder to convert back to a normal array.

> I don't have a problem with `catvalue_type`, let's just make sure we're consistent with `leveltype` regarding the use of underscores.

Well, it's a julian style to minimize the use of underscores in the API, so I would prefer `catvaluetype()`.

nalimilan commented:
> Ideally the tests shouldn't be touched as long as there's no API/behaviour change. It's also a good test of API completeness.

Well, that's maybe your ideal, but that's not mine! It's good to test for API completeness, but it's also good to check that the implementation works as it's supposed to. We do that all the time, including in Julia Base.

> That's the way to fetch T from CategoricalValue{T,R}, it was a minor problem when there was just CategoricalValue, but with an addition of CategoricalString there should be a uniform way to get it from any categorical value [type]. If you don't know T, it makes it e.g. harder to convert back to a normal array.

Right. Currently the only solution is `eltype(levels(x))`. Let's also support `eltype` on `CatValue`, it cannot hurt. `CategoricalPool` can remain as-is.

> Well, it's a julian style to minimize the use of underscores in the API, so I would prefer catvaluetype().

Agreed.

alyst commented:
So I've rebased to the updated master, renamed `valtype()` to `leveltype()` and `[unwrap_]catvalue_type()` to `[unwrap_]catvaluetype()`.
I've also added `leveltype(CategoricalArray)` and `catvaluetype(CategoricalArray)`, because that would be useful.
`catvaluetype()` and `leveltype()` are not exported yet.

nalimilan commented on src/nullablearray.jl:
```diff
@@ -3,39 +3,40 @@ import Base: convert, getindex, setindex!, similar, in
 
-CategoricalArray{Union{T, Null}, N, R}(dims::NTuple{N,Int}; ordered=false) where {T, N, R} =
-    CategoricalArray{Union{T, Null}, N, R}(zeros(R, dims), CategoricalPool{T, R}(ordered))
+# FIXME remove these ctors? (handled by generic CategoricalArray ctors) 
```
Remove (including the commented out ones below).

nalimilan commented on src/nullablearray.jl:
```diff
@@ -3,40 +3,6 @@ import Base: convert, getindex, setindex!, similar, in
-CategoricalMatrix(A::AbstractMatrix{T}; ordered=_isordered(A)) where {T>:Null} =
-    CategoricalMatrix{T}(A, ordered=ordered)
-
 function convert(::Type{CategoricalArray{T, N, R}}, A::AbstractArray{S, N}) where {S, T>:Null, N, R}
```
I wonder whether this constructor is still needed too now that you take care of unwrapping in the low-level constructor.

nalimilan commented on src/pool.jl:
```diff
@@ -3,13 +3,17 @@ function CategoricalPool(index::Vector{S},
+CategoricalPool{T, R, C}(ordered::Bool=false) where {T, R, C} =
+    CategoricalPool{T, R, C}(T[], Dict{T, R}(), R[], ordered)
 CategoricalPool{T, R}(ordered::Bool=false) where {T, R} =
     CategoricalPool(T[], Dict{T, R}(), R[], ordered)
```
Why not change this function just like the one below?

alyst replied:
The one above calls inner ctor directly since `C` is explicitly specified.
The one below calls the one in question.
The one in question calls the first ctor in `pool.jl`, which handles `C` selection.
So I have to repeat `T[], Dict{T,R}...` because they are passed to 2 different ctors.

nalimilan replied:
Actually, I would have kept the method below as it was.

nalimilan commented on src/array.jl:
```diff
@@ -520,6 +498,12 @@ end
+catvaluetype(A::CategoricalArray) = catvaluetype(typeof(A))
+
+leveltype(::Type{T}) where {T <: CategoricalArray} = leveltype(catvaluetype(T))
+leveltype(A::CategoricalArray) = leveltype(typeof(A))
```
AFAICT this method is redundant with the general fallback.

nalimilan commented on src/value.jl:
```diff
@@ -1,145 +1,174 @@
+    CategoricalValue{T, R}
+catvaluetype(::Type{<:AbstractString}, ::Type{R}) where {R} =
+    CategoricalString{R}
+# default to CategoricalString in degenerated case (all nulls)
```
Repeating my comment from above: This sounds a bit weird. Doesn't it work if you return `CategoricalValue{Union{}, R}`? That's really the correct type, since we know it won't contain anything.

alyst replied:
I made it consistent with the current behaviour that picks `CategoricalString` for empty arrays with unspecified type:
```julia
CategoricalArray(dims::Int...; ordered=false) =
    CategoricalArray{String}(dims, ordered=ordered)
```

If for all-nulls case we have `CategoricalValue{Union{}}` then for consistency we should have it for empty array as well.

> since we know it won't contain anything.

The type -- yes, but the array could be modified.

nalimilan replied:
The default constructor for empty arrays is just a convenience to avoid writing `{String}` explicitly since that's the most common use case. It's really no more than that.

OTC, if the user created an array from an `AbstractArray{Null}`, the array is not supposed to be able to contain anything other than `null`. That's important for example if a column of nulls is concatenated later with a column of non-null values, which could be of any type `T`: the column of nulls should not interfere with type promotion, i.e. `T` should be preserved. That's how it works for e.g. `[null]` already. (That's not just theoretical: I've precisely hit this problem in R a few minutes ago with CSV files containing all-null columns which are considered as being boolean.)

alyst replied:
> The default constructor for empty arrays is just a convenience to avoid writing `{String}`

But it also leads to an "interesting" behaviour of e.g. `vcat()` (master, but the same holds for this PR):
```
julia> x = CategoricalVector(0)
0-element CategoricalArrays.CategoricalArray{String,1,UInt32}

julia> y = CategoricalVector{Int}(0)
0-element CategoricalArrays.CategoricalArray{Int64,1,UInt32}

julia> z = CategoricalVector([1])
1-element CategoricalArrays.CategoricalArray{Int64,1,UInt32}:
 1

julia> vcat(x, y)
0-element CategoricalArrays.CategoricalArray{Union{Nulls.Null, String},1,UInt32}

julia> vcat(y)
0-element CategoricalArrays.CategoricalArray{Int64,1,UInt32}

julia> vcat(x, z)
1-element CategoricalArrays.CategoricalArray{Union{Int64, Nulls.Null},1,UInt32}:
 1
```

So the result of `vcat()` depends on the emptyness/nonemptyness of its arguments, just like R depends on all-nulls/not all-nulls.
I guess `vcat()` was tuned at some point to workaround the default type of an empty array.

nalimilan replied:
Good catch. That's a bug with `CategoricalArrays.mergelevels(false, ["a"], Int[])`, which has a fast path for when a set of levels is empty. We should always respect the original element type even if the array is empty.

nalimilan replied:
See https://github.com/JuliaData/CategoricalArrays.jl/pull/89.

nalimilan commented on src/value.jl:
```diff
@@ -1,145 +1,174 @@
+    C(convert(R, level), pool)
+
+Base.convert(::Type{T}, x::T) where {T <: CatValue} = x
+Base.convert(::Type{Union{T, Null}}, x::T) where {T <: CatValue} = x
```
Why is this second definition needed? Also better put these with other `convert` methods.

alyst replied:
I will put, but I thought those were for resolving ambiguities with Base as your comment states.

alyst replied:
`Base.convert(::Type{Union{T, Null}}, x::T) where {T <: CatValue} = x` is an interesting case, it's required to make `x2 == collect(x)` test to pass.
`Base.collect(CategoricalArray{Union{T,Null}})` first generates `Array{Union{Null,CategoricalValue{T}}}` and then fills it with `setindex!(::Array{Union{Null,CategoricalValue{T}}}, ::CategoricalValue{T}, ::Int)`.
So `setindex!()` calls `convert(Union{Null,CategoricalValue{T}}}, ::CategoricalValue{T})`. Normally, it should be handled by `Nulls` generic `convert`, but for categorical values there's fallback convert `Base.convert(::Type{S}, x::CatValue) where {S} = convert(S, get(x))` that seemingly overrides the Null's one as more specific (why?).
So I had to define the one in question.

I guess the fact that `collect()` returns `Array{CategoricalValue}` is not nice, and there should be `collect(CategoricalArray)` and `collect(SubArray{CategoricalArray})`.
But it should be a topic of another PR.

nalimilan replied:
Specificity issues are quite hard to get right (and there may have been some bugs)... And yes, we should probably override `collect`.

nalimilan commented on src/value.jl:
```diff
@@ -1,145 +1,174 @@
+catvaluetype(::Type{T}, ::Type{R}) where {T >: Null, R} =
+    catvaluetype(Nulls.T(T), R)
+catvaluetype(::Type{T}, ::Type{R}) where {T <: CatValue, R} =
+    reftype(T) === R ? T : catvaluetype(leveltype(T), R)
```
Repeating my comment from above: Still the same  remark about `reftype(T) === R`.

alyst replied:
This is correct now and this would be correct if categorical values are extended (without this check there could be unnecessary copies / unexpected changes of behaviour).
I suggest to keep it as is.

nalimilan replied:
But in the even categorical values were extended, we wouldn't want the return type to change depending on whether `R` is the same or not. Imagine `CategoricalString` was only provided by an external package: would you like to have `catvaluetype(CategoricalString{UInt8}, UInt8) == CategoricalString{UInt8}` but `catvaluetype(CategoricalString{UInt8}, UInt32) == CategoricalValue{String, UInt32}`? Custom categorical value types would have to override `catvaluetype` anyway. So better make it very clear, rather than having a different behavior only in a very specific case which is less likely to be spotted.

alyst replied:
> Custom categorical value types would have to override `catvaluetype()` anyway. 

Exactly (otherwise they won't be used), but not the `T<:CatValue` method.
So there should be no situation you describe.

Though the situation that I'm referring to (multiple categorical value implementations for the same level type that are selected depending on the context) are not supported by the current API.
So I'll drop `R` check to make everyone happy.

nalimilan commented on src/typedefs.jl:
```diff
@@ -2,9 +2,10 @@ const DefaultRefType = UInt32
+# Type params:
+# * `T` type of categorized values
+# * `R` integer type for referencing category levels
+# * `V` categorical value type that implements "categorical value" trait
```
Repeating my comment from above: Could you remove quotes around mentions of "categorical value"? I don't find it very useful, and for example in the docstring below they are not present.

nalimilan commented:
Thanks for finishing this!

alyst commented:
Thanks for patient and rigorous reviews :)

nalimilan commented:
I forgot to ask you to update the docs. Would you make a PR to mention quickly the existence of both `CategoricalString <: AbstractString` and `CategoricalValue`? Thanks!
